lib/AST/ASTImporter.cpp

```diff
@@ -4659,19 +4673,20 @@ Decl *ASTNodeImporter::VisitFunctionTemplateDecl(FunctionTemplateDecl *D) {
   if (!Params)
     return nullptr;
 
-  FunctionDecl *TemplatedFD =
-      cast_or_null<FunctionDecl>(Importer.Import(D->getTemplatedDecl()));
-  if (!TemplatedFD)
-    return nullptr;
+  // Check recursive import.
```

Comment:
This is the second solution that tries to avoid the "loop" at template import. Always import the "templated function" first. If the import is called from the templated function (this is indicated by the condition that `TemplatedFD` is non-null) import really the template. If the import is called not from the templated function (`TemplatedFD` should be null) import the templated function, this causes a recursive call with the previous condition, template is imported, and this call returns the already imported template.

But currently the first solution is the likely one.